ooniprobe/TestClasses/MKNetworkTest.mm

```diff
@@ -3,132 +3,258 @@
 #import "VersionUtility.h"
 #import "ReachabilityManager.h"
 
+static NSDictionary *wait_for_next_event(mk_unique_task &taskp) {
+    mk_unique_event eventp{mk_task_wait_for_next_event(taskp.get())};
+    if (!eventp) {
+        NSLog(@"Cannot extract event");
+        return nil;
+    }
+    const char *s = mk_event_serialize(eventp.get());
+    if (s == nullptr) {
+        NSLog(@"Cannot serialize event");
+        return nil;
+    }
+    // Here it's important to specify freeWhenDone because we control
+    // the lifecycle of the data object using `eventp`.
+    NSData *data = [NSData dataWithBytesNoCopy:(void *)s length:strlen(s)
+                                  freeWhenDone:NO];
+    NSError *error = nil;
+    NSDictionary *evinfo = [NSJSONSerialization JSONObjectWithData:data
+                                                           options:0 error:&error];
+    if (error != nil) {
+        NSLog(@"Cannot parse serialized JSON event");
+        return nil;
+    }
+    return evinfo;
+}
+
 @implementation MKNetworkTest
 
 -(id) init {
     self = [super init];
     if (self) {
         self.backgroundTask = UIBackgroundTaskInvalid;
-        [self createMeasurementObject];
+        self.measurements = [[NSMutableDictionary alloc] init];
+        self.settings = [Settings new];
+        //self.settings.log_filepath = [TestUtility getFileNamed:[self.result getLogFile:self.name]];
     }
     return self;
 }
 
-- (void)createMeasurementObject{
-    self.measurement = [Measurement new];
-    //If needed for the second measurement object (websites)
+- (Measurement*)createMeasurementObject{
+    Measurement *measurement = [Measurement new];
     if (self.result != NULL)
-        [self.measurement setResult_id:self.result];
+        [measurement setResult_id:self.result];
     if (self.name != NULL)
-        [self.measurement setTest_name:self.name];
-    [self.measurement save];
+        [measurement setTest_name:self.name];
+    if (self.reportId != NULL)
+        [measurement setReport_id:self.reportId];
+    if (self.network != NULL)
+        [measurement setNetwork_id:self.network];
+    [measurement save];
+    return measurement;
 }
 
-- (void)setResultOfMeasurement:(Result *)result{
-    self.result = result;
-    [self.measurement setResult_id:self.result];
+-(void)runTest{
+    self.backgroundTask = [[UIApplication sharedApplication] beginBackgroundTaskWithExpirationHandler:^{
+        [[UIApplication sharedApplication] endBackgroundTask:self.backgroundTask];
+        self.backgroundTask = UIBackgroundTaskInvalid;
+    }];
+    dispatch_async(
+                   dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+                       NSLog(@"%@", [self.settings getSerializedSettings]);
+                       mk_unique_task taskp{mk_nettest_start([[self.settings getSerializedSettings] UTF8String])};
+                       while (!mk_task_is_done(taskp.get())) {
```

Comment:
@bassosimone you might want to quickly check this piece (until row 154)

Reply:
sure!